config_utilities/demos/demo_config.cpp

```diff
@@ -160,78 +161,362 @@ void declare_config(SubSubConfig& config) {
+    //TODO: check registered
+    std::shared_ptr<ReconfigureInterface> interface = server.registered_interface_.at(typeid(Interface));
+    assert(interface != nullptr);
+
+    std::cout << "Loaded inteface " << typeid(Interface).name() << std::endl;
+    //make new paramter interface to update this config
+    auto handler = std::make_unique<ParameterEvenHandler>(config, key);
+    //add to interface
+    interface->monitorParam(std::move(handler));
   }
 
-  // ======================================== Read the config from file ========================================
-  std::cout << "\n\n----- Reading the config from file -----\n\n" << std::endl;
 
-  // Read the config from file.
-  config = config::fromYamlFile<demo::MyConfig>(my_root_path + "params.yaml");
 
-  std::cout << "Read values i='" << config.i << "', s='" << config.s << "', distance='" << config.distance
-            << "' from file." << std::endl;
-  std::cout << "Enum 'config.my_enum' is now B: " << (config.my_enum == demo::MyConfig::MyEnum::kB) << std::endl;
+private:
+  DynamicReconfigureServer() = default;
+  //map holding the interfaces and their derived type index so that when a config
+  //is registered we can get the right interface
+  //this relies on the type index being the same, maybe worth doing some actual casting to check
+  std::map<std::type_index, std::shared_ptr<ReconfigureInterface>> registered_interface_;
 
-  // Any errors parsing configs will print verbose warnings if desired and use the default values.
-  invalid_config = config::fromYamlFile<demo::MyConfig>(my_root_path + "invalid_params.yaml");
+};
 
-  // ======================================== Printing configs to string ========================================
-  std::cout << "\n\n----- Printing configs to string -----\n\n" << std::endl;
+int main(int argc, char** argv) {
+  if (argc < 2) {
+    std::cerr << "invalid usage! expected resource directory as argument" << std::endl;
+    return 1;
+  }
 
-  // Easier automatic printing of all configs with unit and additional information can be done using the toString():
-  const std::string config_as_string = config::toString(config);
-  std::cout << config_as_string << std::endl;
+  const std::string my_root_path = std::string(argv[1]) + "/";
 
-  // Inclunding "printing.h" also implements the ostream operator for decared config types. The above is thus equivalent
-  // to:
-  std::cout << config << std::endl;
+  // GLobal settings can be set at runtime to change the behavior and presentation of configs.
+  config::Settings().inline_subconfig_field_names = true;
+  config::Settings().print_warnings = true;
+
+  SimpleConfig config;
+  std::cout << config::toString(config) << std::endl;
+  auto interface = std::make_shared<ManualInterface>();
+  DynamicReconfigureServer::registerInterface(interface);
+
+  //I am unclear about this key. Does it need to be namespaced according to anything else in the system?
```

Comment:
the key would be a an identifier that is unique to each server (in the case of a ROS node e.g. resolved to the node). The a global identifier is required so that outside clients can send their updates to the right address. If you run our ROS example you could e.g. have several ROS nodes with dynamic config servers, and in the UI you can select the server + the config for each server to target. For other applications the end user would be responsible to namespace or otherwise disambiguate configs correctly (e.g. by having an ID counter in objects using dynamic configs).